src/core/shutdown.c

```diff
@@ -158,6 +158,41 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+static int sync_with_timeout(void) {
+        sigset_t mask, oldmask;
+        pid_t pid;
+        int r;
+
+        log_info("Syncing filesystems.");
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+        assert_se(sigprocmask(SIG_BLOCK, &mask, &oldmask) == 0);
+
+        /* Due to the possiblity of a remount operation hanging, we
+         * fork a child process and set a timeout. If the timeout
+         * lapses, the assumption is that that particular remount
+         * failed. */
+        pid = fork();
+        if (pid < 0)
+                return log_error_errno(errno, "Failed to fork: %m");
+
+        if (pid == 0) {
+                /* Start the mount operation here in the child */
+                sync();
+                _exit(EXIT_SUCCESS);
+        }
+
+        r = wait_for_terminate_with_timeout(pid, LONG_TIMEOUT_USEC);
+        if (r < 0) {
+                log_notice("Timed out, issuing SIGKILL to "PID_FMT".", pid);
+                (void) kill(pid, SIGKILL);
```

Comment:
If we have a reason this should not be removed (i.e. is required), perhaps we need to comment it.  Thinking of MNT_FORCE, which was removed because readers didn't know what MNT_FORCE was.  (The man page documenting MNT_FORCE was improved following that).

I imagine this process could hold the rootfs open, after exec() of `/run/initramfs/shutdown`.  That doesn't necessarily involve a failure to unmount/remount any filesystem; I think it could be an issue with a block device...

`<aside>`huh, surely dirty [cache pages on a block device](https://lwn.net/Articles/736534/) being another reason we absolutely need to keep sync()?  Can we please add a comment for that, perhaps where we call this function?

Plus, that would imply we need a LOG_ERROR, for the possible loss of unsynced data.  And a message to match.

`</aside>`

So if the `sync()` thread does not die quickly enough, we can expect the initramfs to fail unmounting the root, which is the whole reason for shutdown-to-initramfs.  Right?  If there's nothing more we can do _or log_ about it, then that probably deserves a comment saying so.

Reply:
I can add a log_error notification that the sync failed, and a short comment warning that the sync should not be removed, despite it looking unnecessary.

Reply:
* I would change the message at least to start "Sync timed out".  I at least think it would be best to be 100% clear about what timed out.  "Timed out, issuing SIGKILL to <X>" sounds like it might be something else.
* I also just noticed it looks like "issuing SIGKILL to 99".  git grep says we tend to write "PID 99" or "process 99".

Apologies for rambling.  You acknowledged one of the points I made above, but not the other.   I will try to state them again without mixing them up confusingly.

Using log_error() for the timeout message sounds safer to me, thanks.  Even if I'm struggling a bit thinking of practical scenarios.

* I suggested adding a comment, but after thinking further I would change the "Syncing filesystems" message to "Syncing filesystems and block devices".  That would more accurately describe what's happening.  It makes less potential for confusion, when people are trying to track down a problem.

   sync() can be seen as part of traditional shutdown sequences going back a long way.  I think it's most likely that some people have come to rely on it.  Syncing block devices is not redundant with umount_all(), so this message would point out one reason not to remove sync().  (Apparently it doesn't cover the reason that caused sync() to be re-added.  But I wouldn't expect you to write that up for this PR).

   An alternative message would avoid mentioning filesystems at all, and a comment could explain about covering both filesystems (potentially not backed by a block device) and block devices.  I just don't have a concrete suggestion for that.

* What I was writing about to start with was not the need for sync().  The code I was commenting on was the need to kill() the child process.  The code looked suspicious to me, because it kills the child process, but it doesn't seem to care about what the child does afterwards.  It doesn't seem to wait, for example, for the child to die and release a resource that it holds.

   So I'm wondering whether the kill() does anything useful.

   I notice that the kill() is potentially required, in order to close all FDs of the root filesystem so that it can be unmounted.  Unmounting the root filesystem being the point of running `/run/initramfs/shutdown`.

   If we think the kill() is not going to do anything reliable to a stuck sync() - which is my current guess based on unscientific, not entirely relevant tests?  Then we should consider what the result of that is, and the code should show that.  Even just `/* the child process might be totally stuck; we don't even try to wait for it */`.  Without that, this sticks out as a distraction when you're reading the code.

   This was one of the fuzzy reasons it's useful to log an error message for sync() timeouts, not just a notice.  If the expectation is the process will stay stuck and cause problems later on, I strongly favour marking this as an error when it happens.

   If we think the kill() is important, we shouldn't mind going to sleep for at least a second, to be clear that the kernel schedules the child task to process the SIGKILL.